pkg/status/status.go

```diff
@@ -1,65 +1,89 @@
+	// to report a working operator
+	successRatio           = 1.0
+	maxSyncsBeforeTruncate = 10000
+	syncTruncateValue      = 100
+)
+
+// SyncCh is used to report a sync event
+// The ratio of Successful Syncs / Total Syncs will be used to report operator health to the CVO
+var syncCh = make(chan error)
+
+// status will be a singleton
+var instance *status
+var mu sync.Mutex
 
 type status struct {
 	configClient    *configclient.ConfigV1Client
 	coAPINotPresent bool
 	namespace       string
 	clusterOperator *configv1.ClusterOperator
 	version         string
+	successfulSyncs int
+	syncs           int
 }
 
-// Status is the interface to report the health of the operator to CVO
-type Status interface {
-	SetFailing(message string)
-	SetAvailable(message string)
-	SetProgressing(message string)
-	GetVersion() string
-}
-
-func (s *status) GetVersion() string {
-	return s.version
+// SendSyncMessage is used to send messages to the SyncCh
+func SendSyncMessage(err error) {
+	if instance != nil {
+		// If the coAPI is not available do not send messages to the sync channel
+		if !instance.coAPINotPresent {
+			syncCh <- nil
+		}
+	}
 }
 
-// New returns an initialized Status
-func New(cfg *rest.Config, mgr manager.Manager, namespace string, version string) Status {
-	// The default client serves read requests from the cache which contains
-	// objects only from the namespace the operator is watching. Given we need
-	// to query CRDs which are cluster wide, we add the CRDs to the manager's
-	// scheme and create our own client.
-	if err := v1beta1.AddToScheme(mgr.GetScheme()); err != nil {
-		log.Fatal(err)
+// StartReporting ensures that a status object exists and then starts updating Marketplace's clusteroperator
+func StartReporting(cfg *rest.Config, mgr manager.Manager, namespace string, version string, stopCh <-chan struct{}) {
+	if instance == nil {
+		// status is a singleton and should prevent threads from instantiating multiple instances.
+		mu.Lock()
+		defer mu.Unlock()
+		if instance == nil {
+			instance = new(cfg, mgr, namespace, version)
+			go instance.monitorClusterStatus(stopCh)
+		}
 	}
+}
 
-	controllerClient, err := client.New(cfg, client.Options{Scheme: mgr.GetScheme()})
+// New returns an initialized status
+func new(cfg *rest.Config, mgr manager.Manager, namespace string, version string) *status {
+	// Check if the ClusterOperator API is present on the cluster. If the API
+	// is not present on the cluster set the status.coAPINotPresent to true so
+	// status is not reported
+	k8sInterface, err := kubernetes.NewForConfig(cfg)
 	if err != nil {
 		log.Fatal(err)
 	}
-
-	// Check if the ClusterOperator API is present on the cluster. This is so
-	// that we can continue to work with Kubernetes clusters that don't have
-	// this API.
-	key := client.ObjectKey{
-		Name: "clusteroperators.config.openshift.io",
-	}
-	err = controllerClient.Get(context.TODO(), key, (&v1beta1.CustomResourceDefinition{}))
```

Comment:
With this change, we are no longer are making `get` CRD requests. Please update the `ClusterRole`.

Reply:
Updated.